### PR TITLE
fix(38388) Add support for `this` parameters in iterationTypes and fix caching issues

### LINKED ISSUE #38388: Implicit Symbol.iterator call in for..of loops and object spreading doesn't infer type parameter(s)
**TypeScript Version:**  4.0.0-dev.20200507

**Search Terms:** Symbol.iterator type parameter implicit calling for..of loops object spread destructuring this generic

**Code**

What follows is 3 different ways of iterating over an object using a custom-made iterator. However, using `Symbol.iterator` implicitly via the for..of loop fails to produce the same behavior:

```ts
const obj = {
	x: 1,
	y: 2,
	z: 3,
	*[Symbol.iterator]<T extends object>(
		this: T,
	): Generator<NonNullable<{ [K in keyof T]: [K, NonNullable<T[K]>] }[keyof T]>, void, unknown> {
		for (const entry of Object.entries(this)) yield entry as never;
	},
};

{
	const iter = obj[Symbol.iterator]();

	for (let result = iter.next(); !result.done; result = iter.next()) {
		const { value } = result;
		value; // ["x", number] | ["y", number] | ["z", number]
	}
}

for (const value of obj[Symbol.iterator]()) {
	value; // ["x", number] | ["y", number] | ["z", number]
}

for (const value of obj) {
	value; // NonNullable<{ [K in keyof T]: [K, NonNullable<T[K]>]; }[keyof T]>
	// bad! This should be ["x", number] | ["y", number] | ["z", number]
}

const values = [...obj]; // Array<NonNullable<{ [K in keyof T]: [K, NonNullable<T[K]>]; }[keyof T]>>
// bad! This should be Array<["x", number] | ["y", number] | ["z", number]>
```

**Expected behavior:** `for (const value of obj)` should implicitly do the type equivalent of calling our iterator symbol, i.e. `for (const value of obj[Symbol.iterator]())`. That means our type parameter should be correctly inferred like so: `obj[Symbol.iterator]<T>(this: T)`.

**Actual behavior:** `T` cannot be inferred as the local `this` type, so all derivative types cannot be evaluated and the type stays as its generic version.

[**Playground Link**](https://www.typescriptlang.org/play/index.html?ts=4.0.0-dev.20200506#code/MYewdgzgLgBCBGArGBeGBvAsAKAJAA8AuGARgBodcBPYgJgrwC9iBmB3AKgG0BlKgW3ggANgDoAllACmAJwCGUEDIC6AHgAqMKfmlgAJhDhIpwKAD4AFJVxQAFuIjF17AJTEA4lLCyFS1QDlwfwBXYWE5eGEpVXQYLgBpGHEwGABrKSoQADMYdWViBLIYQLAQsIiojQTlM2UYAF8udMycvLMigDcQcT0i4LBUsBAAdzAzDGsspRgLUEhYLygZKjgcgHljU1FFmXEpCAs7BxcXGCo94T0tMCWVuUNvDtkAbkp6hnrX7BwsPDnoJLSGSoIyIXgCIRiSQ+RQqCwuL6UKbAixRWAyfahWBoaEyUTeHTw54wACEGIgWNEenAUmJ5KxINx+O0UHhp1+uFw-1gsQ6cmEwSkDRB9OEUC+nL5AtpMAA9LK4gAifCKopgYKCWR1AA+Sqoqpg6s1Khguq4isYBqN8C1bxw9RwOGRM25MClgtWoPBghEEiBvjhJwmeHdMvlSpVao1NpNZsV+qjxp1SstiZjyntjuwztm4ABoc9CEQ7MooeJ4ZKZXCkWisQSSRSzWyuXycXiRUroWrlXU1VqxMaTdaNUo4fgcj0JNy9kMEFsIFCVxtEat0a1pr1q6TG-NqcNa5Ume+2FdocMaC4oivReU5YVAEEZPIqAEgl2KrW2w20hlm3kCu2xRvuUNZVPENS3g0TS-sOZhmDgY4TlO6gzjAc4LpcMDLo+z6qOakb7tucYJoR6Y7haW7pmYQA) 

**Related Issues:**  n/a

### LINKED ISSUE #39693: for-of loop with intersection of array types produces a union of element types
**TypeScript Version:**  3.9.2, 4.0.0-beta

**Search Terms:** intersection, for-of, union, iterable, iterator, iterate, array

**Expected behavior:**
When you use a `for..of` loop to iterate over the elements of an intersection of arrays (or maybe other iterables), what type should the elements be?  I would expect either:
* you get the same type as when you index into the array: an intersection of the element types; or
* you get the same type as when you use the iterator method manually: the first element type because the iterator methods are overloads.

**Actual behavior:**
Iterating over an intersection of arrays with a `for..of` loop produces a *union* of their element types for some reason.

**Related Issues:**
#11961: intersection of array types results in overloaded methods (this would maybe imply overloaded iterators, but that's not happening here)

**Aside:**
I'm not sure why you'd want an intersection of array types in the first place; but the behavior showed up in [a Stack Overflow question](https://stackoverflow.com/questions/63010203/typescript-intersection-fails-on-list-for-loop) and I'm at a loss understanding why we get a union here.

**Code**
```ts
declare const arr: Array<{ a: string }> & Array<{ b: number }>;

for (const elemItr of arr) {
  // { a: string } | {b : number } 😕
  elemItr.a.toUpperCase(); // error!
  elemItr.b.toFixed(); // error!
}

// I expected either this (intersection of element types)
const elemIdx = arr[0]; // { a: string; } & { b: number; }
elemIdx.a.toUpperCase(); // okay
elemIdx.b.toFixed(); // okay

// or this (overloaded iterators giving the first element type only)
const iter = arr[Symbol.iterator];
/* const iter: {
    (): IterableIterator<{ a: string }>;
    (): IterableIterator<{ a: string }>;
} & {
    (): IterableIterator<{ b: number }>;
    (): IterableIterator<{ b: number }>;
} */
const result = arr[Symbol.iterator]().next();
if (!result.done) {
  result.value.a.toUpperCase(); // okay
  result.value.b.toFixed(); // error, expected this 
}
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
for (const elemItr of arr) {
    // { a: string } | {b : number } 😕
    elemItr.a.toUpperCase(); // error!
    elemItr.b.toFixed(); // error!
}
// I expected either this (intersection of element types)
const elemIdx = arr[0]; // { a: string; } & { b: number; }
elemIdx.a.toUpperCase(); // okay
elemIdx.b.toFixed(); // okay
// or this (overloaded iterators giving the first element type only)
const iter = arr[Symbol.iterator];
/* const iter: {
    (): IterableIterator<{ a: string }>;
    (): IterableIterator<{ a: string }>;
} & {
    (): IterableIterator<{ b: number }>;
    (): IterableIterator<{ b: number }>;
} */
const result = arr[Symbol.iterator]().next();
if (!result.done) {
    result.value.a.toUpperCase(); // okay
    result.value.b.toFixed(); // error, expected this 
}
```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?ts=4.0.0-beta&ssl=8&ssc=1&pln=3&pc=1#code/CYUwxgNghgTiAEYD2A7AzgF3rGAueAgjDFAJ4A8A3tvpjAJYoDm8AvgHzwBkhxZV8AEb4UAVwC2gkDDbsA3ACgFAMyQyAFMnRYQEEOICSGGUmXZiASniUF8eAHp71mvDqMWreAB9rg+CIkpGU9AXg3AVD3beF19IxgAOig4jCQAVQAHNOkAYSg0EHULOQcnaRg1AEJI6MNjOMEkpAAxegAPEGACosco4gqFTyVugyiWzLAMdqj6DAALaXhZ+jR4dUYJmDzx+lR4Uyi9cRAULAxSTLQLBS1MfZjgFvgAXnMYAG0ABgBdLqdqKFpjO4ip4eNRhPAxJJpMCFNUDPcEg10pkYDk8p1irsANZkWEHeEtOoNZptDqFTFIHGkQZONQLGZLFZIABu0ggSCgoGA8Gm0igyQ28CY9GZ7npCGU9A2OgORxOZwQqAgpEu1ywvJkzxwrwAyqRJEgIHENfy1N8FPYAFSIVA3DX4Gx2OwFfBGPmCPRukgCgT-VyA5iyRROlYWV3rKAekBe00wX0AhiBjiKEHWSLOsPwGNRmM+sEBKHBeTp0Ph92eiN5oQFoJB-rwS32K62rBwNCiCBYLXEXX6wSG42Vs0FOIoEAtDCdBT0MzqcptjsYOLAVAgKyO+ALztxZlQCCiECI5LI7K5fLk7qU3F2LdL3f7w-1ZIk9oY7qlNQAGhGYwm3MWyz9AoQA)

### PR DESCRIPTION
**Pull request by Validark**
Fixes #38388
Fixes #39693

This PR:
  - Adds support for `this` parameters in iterationTypes!
```ts
const obj = {
	x: 1,
	y: 2,
	z: 3,

	*[Symbol.iterator]<P>(this: P) {
		for (const k in this) yield k;
	},
};

for (const value of obj) {
    console.log(value); // value is `"x" | "y" | "z"`
}
```
- Fixes the iterator caching logic. Looking at the code, I noticed a number of problems that could arise from the caching logic. [Here is a reproduction of two of the issues I found](https://www.typescriptlang.org/play?target=99&ts=4.1.0-beta#code/GYVwdgxgLglg9mABMOcAUcBGArAXAKAEgBvRAD10QCIyqAaRAbQGUBPAW0zgBsA6GKAFMATgEMocYQF00ASkoBxQWBHjJAHlKZKwUdwDOgxAF8AfCcQAfIqVaUqrKifyzExIimGI0EBPqiIANaIcMAhOK7uhIQAbnAwACZBANxExvgekt6+YP5BIWFY2JGIAPSliPoAFnAg3EkiwlmYIAEJcIL6YADkAZiCEKIghogC3fqIgxBVgglEsfFJgamE6en4oJCwCIhQrAAOggCC+obCUOoAKqZoFIiXJXGJ5MnO+KL6rJDI4NDwYAAqZAYHCUKIsDhcPgCVQSaRyRTKWEafzCGBgADmDCeCQY4ECYDgAHcwKYVhDODxeB8vhAAJJCMRwmTyRAnWlKFRMjRgECcERk-DGSKZLw+PwBVgFcLFNzzPaHE5nC6o9EYm6sWQrdaETyIURE0QCbISxBS0IykXRBXHU4iC68-nCDVa-CId0e93le7MXZVdGBCYAA1YQdGE1ElSgaMxiH6g2GRjGEymMySRKamO4rDSQoyNO+mz+CCBGPURwYACEbkVKOzIAzVJhuIJG9zhOXzAAyRBt0TN1uMtQd6uyfB1z6QTnIjsVxA4vFgBKCYDo2bmKI5PKMCBSRAAXhlqU93t87HYygCtQCUBm+q4MSM3DXuzglUERlvT5f-W4xMmVSiJiSZQBMNoAIQZIQrAwII9RAkUx4et6lz+hMMATJgHwwIM3DZn6Rj6KIF76mBd79nAj6IIIZDEfsLZxq0iBEgIVSINOaIQPobrIRUhJCH64gEZUxFGCGsHwWGzZwBAwS-v+7SdD0ASNFkMTqXm7yThAPxbP8QJVOWVY1qCbLaX2A59nCnaID2FktlZGijuOZkcki7ZGfOiyLsuq4qAkG5EDBcEJAhOCpMYQA). Basically, `getIterationTypesOfIterable` had 3 problems:
    - Cached iterationTypes don't error like they should (see the first bug in the playground).
      - To fix this, we don't cache iterationTypes that produced errors, which makes `getIterationTypesOfIterableSlow` get called every time and produce the proper errors.
      - We can also take advantage of this design by detecting in `getIterationTypesOfIterable` whether something was cached to determine whether or not to cache or `reportTypeNotIterableError` (which we don't want to do if there was already an error from `resolveCall`, for example).
      - Keep in mind that caches should only be read from or written to when `implicitThisType === type`. 
    - When `use` permits either AsyncIterables or (Sync)Iterables converted via `getAsyncFromSyncIterationTypes` it would call these functions in this order: `getIterationTypesOfIterableCached(type, asyncResolver) || getIterationTypesOfIterableFast(type, asyncResolver) || getIterationTypesOfIterableCached(type, syncResolver) || getIterationTypesOfIterableFast(type, syncResolver) || getIterationTypesOfIterableSlow(type, asyncResolver) || getIterationTypesOfIterableSlow(type, syncResolver)`. This is incorrect because an object could have an `Symbol.asyncIterator` that is neither cached nor an instance of a simple global type, and that logic could then fall back on the `syncResolver` section. Now it runs `cached || fast || slow` in order for both `Symbol.asyncIterator` and `Symbol.iterator` when applicable. This led to the 2nd and 3rd bugs demonstrated in the playground link above.
    - `iterationTypesOfAsyncIterable` was shared between `Symbol.iterator` and `Symbol.asyncIterator`. Now `iterationTypesOfAsyncedSyncIterable` is used to store `Symbol.iterator` calls that got converted via `getAsyncFromSyncIterationTypes`.

To make implicit `Symbol.iterator` calls like actual `Symbol.iterator` calls, I hooked them up to the `resolveCall` function. This function and a few of the ones it calls had to be modified so that arbitrary `Node`s could be passed into it without analyzing the `Node` as whatever it is. The implicit call case is differentiated by the `implicitThisType` being passed around. I.e. when `implicitThisType` is present, then the node should be treated solely as an `errorNode`, and the only NodeKind-specific behavior should be in getting the boundaries for error messages. `typeArguments` and parametric `args` should always be empty.

Implicit iterator calls do have different behavior from regular calls though. Intersected iterator functions will produce intersected results instead of just grabbing the first one. See #39693. Unioned iterator functions will evaluate each unioned constituent separately (passing the constituent, not the top-level type, as the `implicitThisType`) and the results will be unioned together. [See here for examples](https://github.com/Validark/TypeScript/blob/201b55fbb746354eaaf7e93aa584f273bff79f19/tests/cases/compiler/genericIterables.ts). This means there will be inconsistency between how `for (const x of obj)` vs `for (const x of obj[Symbol.iterator]())` is evaluated, although there was already inconsistency before and I think [it is for the better](https://github.com/Validark/TypeScript/blob/201b55fbb746354eaaf7e93aa584f273bff79f19/tests/cases/compiler/genericIterables.ts).

### CODE REVIEW

**Comment by Validark** [comment]
I noticed on my machine that the errors in `errorsInGenericTypeReference` changed from `Could not find symbol V` to `Cannot find name 'V'`. Is that actually the product of one of my changes...? I don't think it is, but I could update the files if wanted.

File: tests/baselines/reference/errorsInGenericTypeReference.errors.txt [hunk 1]
```diff
@@ -1,5 +1,4 @@
 tests/cases/compiler/errorsInGenericTypeReference.ts(11,17): error TS2304: Cannot find name 'V'.
-tests/cases/compiler/errorsInGenericTypeReference.ts(17,31): error TS2304: Cannot find name 'V'.
```
Comment (Validark) [inline tests/baselines/reference/errorsInGenericTypeReference.errors.txt:2]: This is probably not good. I am still pretty sure this has nothing to do with my changes, I just wanted to update the baseline references so my tests don't fail. Perhaps another issue should be opened for this case?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -26267,33 +26268,33 @@ namespace ts {
             return true;
         }
 
-        function callLikeExpressionMayHaveTypeArguments(node: CallLikeExpression): node is CallExpression | NewExpression | TaggedTemplateExpression | JsxOpeningElement {
+        function callLikeExpressionMayHaveTypeArguments(node: Node): node is CallExpression | NewExpression | TaggedTemplateExpression | JsxOpeningElement {
```
Comment (rbuckton) [inline src/compiler/checker.ts:26271]: Why was the precision for this argument's type reduced?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -26501,46 +26512,48 @@ namespace ts {
...
-                    // Create a type mapper for instantiating generic contextual types using the inferences made
-                    // from the return type. We need a separate inference pass here because (a) instantiation of
-                    // the source type uses the outer context's return mapper (which excludes inferences made from
-                    // outer arguments), and (b) we don't want any further inferences going into this context.
-                    const returnContext = createInferenceContext(signature.typeParameters!, signature, context.flags);
-                    const returnSourceType = instantiateType(contextualType, outerContext && outerContext.returnMapper);
-                    inferTypes(returnContext.inferences, returnSourceType, inferenceTargetType);
-                    context.returnMapper = some(returnContext.inferences, hasInferenceCandidates) ? getMapperFromContext(cloneInferredPartOfContext(returnContext)) : undefined;
+        function inferTypeArguments(node: Node | undefined, signature: Signature, args: readonly Expression[], checkMode: CheckMode, context: InferenceContext, implicitThisType?: Type): Type[] {
```
Comment (rbuckton) [inline src/compiler/checker.ts:26515]: Another case of using a less precise type, and I'm still not clear on why `implicitThisType` should cause most of this function to be ignored.

**Review by rbuckton — COMMENTED** [review]
There's a lot going on here and it will take me some time to review. I'm not certain I agree with some of the changes here, as there are a large number of places where you've switched to use a less precise type so that you can treat `node` as only an "error node". A separate `errorNode` argument argument would likely be preferable in those cases.

It might be worth breaking out the changes to fix the caching issues you mentioned into a separate PR, as those changes seem less controversial to me.

**Comment by Validark** [comment]
> There's a lot going on here and it will take me some time to review. I'm not certain I agree with some of the changes here, as there are a large number of places where you've switched to use a less precise type so that you can treat `node` as only an "error node". A separate `errorNode` argument argument would likely be preferable in those cases.
> 
> It might be worth breaking out the changes to fix the caching issues you mentioned into a separate PR, as those changes seem less controversial to me.

Thanks for getting back to me! I really wasn't sure whether to prioritize making fewer changes and making it work with what's already there or to make changes like you suggested. I tried to think of how to solve the problem while making the smallest number of changes, then take one step towards what I would consider ideal. I wasn't sure what kind of philosophy there is about these kinds of things so I just guessed. Since my guess was wrong, I'll go back and make the changes I think would be ideal instead of trying to guess where my hands should be tied. I did kinda deviate from that anyway and got caught up a bit when I noticed a bunch of other issues while looking at the code. I will work on this more this week. I'll make this better.

**Comment by RyanCavanaugh** [comment]
🧹 Closing due to staleness. Please open a new PR if you'd like to continue. Thanks!

**Pull request closed without merging**